test/integration/util.sh

```diff
@@ -0,0 +1,224 @@
+# to run dbus-broker under various tools (or a completely custom-built dbus-broker version) without risking
+# damage to the underlying test machine. The container simply combines the /etc and /usr directories from the
+# host with our own additions using overlayfs, which is then bind-mounted into the container, so we do a full
+# user-space boot without needing to build a custom image or restart the underlying test machine itself.
+#
+# The function exports/modifies three environment variables:
+#   - $CONTAINER_NAME - container name that can be used to identify the machine in machinectl calls (or in
+#                       direct calls to the systemd-nspawn@.service template)
+#   - $CONTAINER_MACHINE_ID - machine ID of the container, which can be used to locate the container's journal
+#                             under /var/log/journal/$CONTAINER_MACHINE_ID
+#   - $CONTAINER_OVERLAY - upper layer of the container overlayfs that can be used to add additional bits into
+#                          the final container (note that only /etc and /usr subdirectores from this direcory
+#                          are used)
+#
+# Once the container is ready, it can be booted up using container_start(). To execute commands inside the
+# container, container_run() and container_run_user() might come in handy.
+container_prepare() {
+    # Export a couple of env variables which can be used to track/alter the container
+    CONTAINER_NAME="dbus-broker-container-$RANDOM"
+    CONTAINER_MACHINE_ID="$(systemd-id128 new)"
+    CONTAINER_OVERLAY="/var/lib/machines/$CONTAINER_NAME"
+
+    # Switch SELinux to permissive (if enabled), so it doesn't interfere with the container shenanigans below.
+    setenforce 0 || :
+    # We need persistent journal for the systemd-nspawn --link= stuff
+    mkdir -p /var/log/journal
+    journalctl --flush
+
+    # Prepare the nspawn container service
+    mkdir -p "/var/lib/machines/$CONTAINER_NAME"
+    # Notes:
+    #   - with systemd v256+ this can be replaced by systemctl edit --stdin --runtime ..., and the
+    #     mkdir/daemon-reload can be dropped
+    #   - systemd-nspawn can't overlay the whole rootfs (/), so we need to cherry-pick a couple of subdirectories
+    #     we're interested in (in this case it's pretty simple, since dbus-broker installs everything under /usr,
+    #     and we need /etc with our dbus-broker.service override)
+    #   - since the whole container is ephemeral, use --link-journal=host, so the journal directory for the
+    #     container is created on the _host_ under /var/log/journal/<machine-id> and bind-mounted into the
+    #     container; that way we can fetch the container journal for debugging even if something goes horribly
+    #     wrong
+    mkdir -p "/run/systemd/system/systemd-nspawn@$CONTAINER_NAME.service.d"
+    cat >"/run/systemd/system/systemd-nspawn@$CONTAINER_NAME.service.d/override.conf" <<EOF
+    [Service]
+# We'll handle the coredumps on the host instead
```

Comment:
It's really cool that it's possible to collect coredumps on the host there. GH Actions are run in containers so coredumps go to the underlying host actions have no access to. I guess it's convincing enough to switch to this testing infrastructure :-)